coordiante_transform/src/coordinate_transform_component.cpp

```diff
@@ -0,0 +1,187 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include <rclcpp/rclcpp.hpp>
+#include <rclcpp_components/register_node_macro.hpp>
+#include <stsl_interfaces/msg/tag_array.hpp>
+#include <visualization_msgs/msg/marker_array.hpp>
+#include <visualization_msgs/msg/marker.hpp>
+#include <geometry_msgs/msg/point.h>
+#include <geometry_msgs/msg/quaternion.h>
+#include <tf2_ros/transform_listener.h>
+#include <tf2_eigen/tf2_eigen.h>
+
+using namespace std::chrono_literals;
+
+namespace coordinate_transform {
+class CoordinateTransformComponent : public rclcpp::Node
+{
+public:
+  explicit CoordinateTransformComponent(const rclcpp::NodeOptions & options)
+  : rclcpp::Node("coordinate_transformer", options), tf_buffer_(get_clock()), tf_listener_(tf_buffer_)
+  {
+    tag_sub_ = this->create_subscription<stsl_interfaces::msg::TagArray>("/aruco_tag_detector/tags",
+            10, std::bind(&CoordinateTransformComponent::DetectionCallback, this, std::placeholders::_1));
+
+    tag_publisher_ = this->create_publisher<stsl_interfaces::msg::TagArray>("~/tags_transformed", 1);
+    visualization_pub_ = this->create_publisher<visualization_msgs::msg::MarkerArray>("~/tags_visual", 1);
+  }
+
+private:
+  tf2_ros::Buffer tf_buffer_;
+  tf2_ros::TransformListener tf_listener_;
+  rclcpp::Subscription<stsl_interfaces::msg::TagArray>::SharedPtr tag_sub_;
+  rclcpp::Publisher<stsl_interfaces::msg::TagArray>::SharedPtr tag_publisher_;
+  rclcpp::Publisher<visualization_msgs::msg::MarkerArray>::SharedPtr visualization_pub_;
+
+  void DetectionCallback(const stsl_interfaces::msg::TagArray::SharedPtr tag_array_msg)
+  {
+    std::string tf_error_string;
+    if (!tf_buffer_.canTransform(
+            "base_footprint", tag_array_msg->header.frame_id,
+            tag_array_msg->header.stamp, tf2::durationFromSec(0.1),
+            &tf_error_string))
+    {
+      RCLCPP_WARN_THROTTLE(
+              get_logger(), *get_clock(), 1000, "Could not lookup transform. %s",
+              tf_error_string);
+      return;
+    }
+
+    // find the transform from the camera to base footprint
+    const auto tf_transform =
+            tf_buffer_.lookupTransform("base_footprint", "camera_link", tag_array_msg->header.stamp);
+    const Eigen::Isometry3d eigen_transform = tf2::transformToEigen(tf_transform);
+
+    // creates a matrix that goes from camera to standard ROS coordinates
+    Eigen::Matrix4d optical = getRotMatForOpticalFrame();
+
+    // BEGIN STUDENT CODE
+    // create a new tag array message
+    stsl_interfaces::msg::TagArray new_tag_array_msg;
+    // copy the header information
+    new_tag_array_msg.header = tag_array_msg->header;
```

Comment:
Header is just frame and stamp now (no more seq), so I'd suggest it may be clearer to just copy the stamp and set the frame_id. (Rather than copying the whole header and re-writing the frame_id).

```suggestion
    new_tag_array_msg.header.stamp = tag_array_msg->header.stamp;
```